gateway/ln-gateway/src/lnd.rs

```diff
@@ -242,16 +243,14 @@ impl GatewayLndClient {
                     // Break if we got a response back from the LND node that indicates the payment
                     // hash was not found.
                     if e.code() == Code::NotFound {
-                        break;
+                        return Ok(None);
```

Comment:
In the code above I think you should check `payment.failure_reason` before returning the preimage, if there is an error the preimage will be invalid

Reply:
Also check `.status` field to see if equals to `SUCCEEDED`

Reply:
Probably sufficient to just check the `status` field, then return the `failure_reason` in the error message?

Reply:
Changed the above to check the `status` field